Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalPipeline.cpp

```diff
@@ -0,0 +1,4398 @@
+    plMetalFragmentShaderDescription description;
+    memset(&description, 0, sizeof(description));
+    description.numLayers = fCurrNumLayers = 1;
+
+    description.Populate(proj, 0);
+    // DX sets the color invert when the final color should be inverted. Not sure why!
+    if (proj->GetBlendFlags() & hsGMatState::kBlendInvertFinalColor) {
+        description.blendModes[0] |= hsGMatState::kBlendInvertColor;
+    }
+
+    plMetalMaterialPassPipelineState      materialShaderState(&fDevice, vRef, description);
+    plMetalDevice::plMetalLinkedPipeline* linkedPipeline = materialShaderState.GetRenderPipelineState();
+
+    fState.fCurrentPipelineState = linkedPipeline->pipelineState;
+    fDevice.CurrentRenderCommandEncoder()->setRenderPipelineState(linkedPipeline->pipelineState);
+    fDevice.CurrentRenderCommandEncoder()->setFragmentTexture(tex->fTexture, 0);
+    MTL::SamplerState* samplerState = fDevice.SampleStateForClampFlags(hsGMatState::hsGMatClampFlags(proj->GetClampFlags()));
+    fDevice.CurrentRenderCommandEncoder()->setFragmentSamplerState(samplerState, 0);
+    fState.layerStates[0].clampFlag = hsGMatState::hsGMatClampFlags(proj->GetClampFlags());
+
+    // Okay, render it already.
+
+    render.RenderPrims();
+}
+
+// IRenderProjectionEach ///////////////////////////////////////////////////////////////////////////////////////
+// Render any lights that are to be projected onto each pass of the object.
+void plMetalPipeline::IRenderProjectionEach(const plRenderPrimFunc& render, hsGMaterial* material, int iPass, const plSpan& span, const plMetalVertexBufferRef* vRef)
+{
+    // If this is a bump map pass, forget it, we've already "done" per-pixel lighting.
+    // if( fLayerState[iPass].fMiscFlags & (hsGMatState::kMiscBumpLayer | hsGMatState::kMiscBumpChans) )
+    //    return;
+
+    // Push the LayerShadowBase override. This sets the blend
+    // to framebuffer as Add/ZNoWrite and AmbientColor = 0.
+    static plLayerLightBase layLightBase;
+
+    // For each projector:
+    int k;
+    for (k = 0; k < fProjEach.size(); k++) {
+        // Push it's projected texture as a piggyback.
+        plLightInfo*              li = fProjEach[k];
+        plMetalMaterialShaderRef* mRef = static_cast<plMetalMaterialShaderRef*>(material->GetDeviceRef());
+
+        plLayerInterface* proj = li->GetProjection();
+        hsAssert(proj, "A projector with no texture to project?");
+        IPushProjPiggyBack(proj);
+
+        // Enable the projecting light only.
+        IEnableLight(0, li);
+        fLights.count = 1;
+
+        AppendLayerInterface(&layLightBase, false);
+
+        IHandleMaterialPass(material, iPass, &span, vRef, false);
+
+        IScaleLight(0, true);
+
+        // Do the render with projection.
+        render.RenderPrims();
+
+        RemoveLayerInterface(&layLightBase, false);
+
+        // Pop it's projected texture off piggyback
```

Comment:
```suggestion
        // Pop its projected texture off piggyback
```

"it is projected texture"